tests/phpunit/includes/build-tree-representation.php

```diff
@@ -0,0 +1,263 @@
+					uasort(
+						$sorted_attributes,
+						static function ( $a, $b ) {
+							$a_has_ns = str_contains( $a, ':' );
+							$b_has_ns = str_contains( $b, ':' );
+
+							// Attributes with `:` should follow all other attributes.
+							if ( $a_has_ns !== $b_has_ns ) {
+								return $a_has_ns ? 1 : -1;
+							}
+
+							$a_has_sp = str_contains( $a, ' ' );
+							$b_has_sp = str_contains( $b, ' ' );
+
+							// Attributes with a namespace ' ' should come after those without.
+							if ( $a_has_sp !== $b_has_sp ) {
+								return $a_has_sp ? 1 : -1;
+							}
+
+							return $a <=> $b;
+						}
+					);
+
+					foreach ( $sorted_attributes as $attribute_name => $display_name ) {
+						$val = $processor->get_attribute( $attribute_name );
+						/*
+							* Attributes with no value are `true` with the HTML API,
+							* we use the empty string value in the tree structure.
+							*/
+						if ( true === $val ) {
+							$val = '';
+						} elseif ( 'class' === $attribute_name ) {
+							$class_names = iterator_to_array( $processor->class_list() );
+							sort( $class_names, SORT_STRING );
+							$val = implode( ' ', $class_names );
+						} elseif ( 'style' === $attribute_name ) {
+							$normalized_style = '';
+							foreach ( explode( ';', $val ) as $style ) {
+								if ( empty( trim( $style ) ) ) {
+									continue;
+								}
+								list( $style_key, $style_val ) = explode( ':', $style );
+
+								$style_key = trim( $style_key );
+								$style_val = trim( $style_val );
+
+								$normalized_style .= "{$style_key}:{$style_val};";
+							}
+							$val = $normalized_style;
+						}
+						$output .= str_repeat( $tree_indent, $tag_indent + 1 ) . "{$display_name}=\"{$val}\"\n";
+					}
+				}
+
+				// Self-contained tags contain their inner contents as modifiable text.
+				$modifiable_text = $processor->get_modifiable_text();
+				if ( '' !== $modifiable_text ) {
+					$output .= str_repeat( $tree_indent, $tag_indent + 1 ) . "\"{$modifiable_text}\"\n";
+				}
+
+				if ( 'html' === $namespace && 'TEMPLATE' === $token_name ) {
+					$output .= str_repeat( $tree_indent, $indent_level ) . "content\n";
+					++$indent_level;
+				}
+
+				break;
+
+			case '#cdata-section':
+			case '#text':
+				$text_content = $processor->get_modifiable_text();
+				if ( '' === trim( $text_content, " \f\t\r\n" ) ) {
+					break;
+				}
+				$was_text = true;
+				if ( '' === $text_node ) {
+					$text_node .= str_repeat( $tree_indent, $indent_level ) . '"';
+				}
+				$text_node .= $text_content;
+				break;
+
+			case '#funky-comment':
+				// Comments must be "<" then "!-- " then the data then " -->".
+				$output .= str_repeat( $tree_indent, $indent_level ) . "<!-- {$processor->get_modifiable_text()} -->\n";
+				break;
+
+			case '#comment':
+				// Comments must be "<" then "!-- " then the data then " -->".
+				$comment = "<!--{$processor->get_full_comment_text()}-->";
+
+				// Maybe the comment is a block delimiter.
```

Comment:
okay this definitely looks like a good case for `scan_delimiters()`, though I did propose at one point exposing the PCRE pattern for finding these, and that could also be helpful here since it grabs opening, closing, and void delimiters in one go

Reply:
I don’t think it would be terrible here to borrow the PCRE pattern in the meantime from the default block parser.

also, it looks like you can actually take this from `WP_Block_Parser` directly

```php
$parser           = new WP_Block_Parser();
$parser->document = $comment;
$parser->offset   = 0;
list( $delimiter_type, $block_name, $attrs, $start_offset, $token_length ) = $parser->next_token();

switch ( $delimiter_type ) {
	case 'no-more-tokens':
	case 'void-block':
	case 'block-closer':
	case 'block-opener':
}
```